src/libs/relay/conduit_relay_zfp.cpp

```diff
@@ -0,0 +1,183 @@
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+// IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+//-----------------------------------------------------------------------------
+///
+/// file: conduit_relay_zfp.cpp
+///
+//-----------------------------------------------------------------------------
+
+#include "conduit_relay_zfp.hpp"
+#include "zfpfactory.h"
+
+//-----------------------------------------------------------------------------
+// -- begin conduit:: --
+//-----------------------------------------------------------------------------
+namespace conduit
+{
+
+//-----------------------------------------------------------------------------
+// -- begin conduit::relay --
+//-----------------------------------------------------------------------------
+namespace relay
+{
+
+
+//-----------------------------------------------------------------------------
+// -- begin conduit::relay::io --
+//-----------------------------------------------------------------------------
+namespace io
+{
+
+zfp::array*
+unwrap_zfparray(const Node &node)
+{
+    Node compressed_data = node.fetch_child(ZFP_COMPRESSED_DATA_FIELD_NAME);
+
+    // verify word size is readable with zfp
+    // zfp's bitstream consists of uint words
+    bool is_readable = true;
+    switch(stream_word_bits) {
+        case 64:
+            is_readable = compressed_data.dtype().is_uint64();
+            break;
+
+        case 32:
+            is_readable = compressed_data.dtype().is_uint32();
+            break;
+
+        case 16:
+            is_readable = compressed_data.dtype().is_uint16();
+            break;
+
+        case 8:
+            is_readable = compressed_data.dtype().is_uint8();
+            break;
+
+        default:
+            is_readable = false;
+            break;
+    }
+
+    if(!is_readable) {
+        return NULL;
+    }
+
+    zfp::array::header header[1];
+    memcpy(header->buffer, node.fetch_child(ZFP_HEADER_FIELD_NAME).data_ptr(), sizeof(header));
+
+    try {
+        return zfp::array::construct(header[0], static_cast<uchar*>(compressed_data.data_ptr()), compressed_data.allocated_bytes());
+    } catch(std::exception const &) {
+        // could be zfp::array::header::exception, or std::bad_alloc
+        return NULL;
+    }
+}
+
+template<typename T>
+void
+cast_and_set_compressed_data(Node &dest, uchar* compressed_data, size_t num_data_words)
+{
+    void* intermediate_ptr = static_cast<void*>(compressed_data);
+    dest[ZFP_COMPRESSED_DATA_FIELD_NAME].set(static_cast<T*>(intermediate_ptr), num_data_words);
+}
+
+int
+wrap_zfparray(const zfp::array *arr,
+         Node &dest)
+{
+    // store header
+    zfp::array::header header;
+    try {
+        header = arr->get_header();
+    } catch(zfp::array::header::exception const &) {
+        return 1;
```

Comment:
Since we're returning an integer to indicate success status for this function, it would probably be better to use an integer other than 1 here to differentiate this case from the bitwidth mismatch case accounted for below.

Reply:
Sure I'll change the return value so we can differentiate error causes